86c730dbp/architecture/deploy-self-hosted-models

ori-small opened the pull request:
added hugging face download request and tracked download, and added cuda and python version to ModelVersion.

AmitMizrachi618 committed:
First commit for huggingface to storage new objects.
First commit for huggingface to storage new objects.

(#86c6e25nn)

AmitMizrachi618 committed:
Adding cuba and python version to Model, editing norman hug requests.
Adding cuba and python version to Model, editing norman hug requests.

(#86c6e25nn)

AmitMizrachi618 committed:
Small changes to huggingface objects, making them operational for a test run.
Small changes to huggingface objects, making them operational for a test run.

(#86c6e25nn)

AmitMizrachi618 committed:
Bugfix to HuggingFaceDownloadRequest which has =str rather than :str.
Bugfix to HuggingFaceDownloadRequest which has =str rather than :str.

(#86c6e25nn)

AmitMizrachi618 committed:
Adding asset_type to ModelAsset class, to differenciate between requirements and inference py files.
Adding asset_type to ModelAsset class, to differenciate between requirements and inference py files.

(#86c6e25nn)

AmitMizrachi618 committed:
Adding HuggingFace and Git to receive format and input type, bug fix where we've had coda written instead of cuda.
Adding HuggingFace and Git to receive format and input type, bug fix where we've had coda written instead of cuda.

(#86c6e25nn)

AmitMizrachi618 committed:
Adding properties to Huggingface download request and tracked download.
Adding properties to Huggingface download request and tracked download.

(#86c6e25nn)

Dolevg2411 committed:
Adding cuda and python version and asset type.

(#86c6e265r)

Dolevg2411 committed:
Rename asset type to asset name.

(#86c6e265r)

Dolevg2411 committed:
Rename asset type to asset name.

(#86c6e265r)

Dolevg2411 committed:
Rename asset type to asset name.

(#86c6e265r)

ori-small committed:
Added addition status flag names according to DB changes.
Added addition status flag names according to DB changes.

(#86c246m16)

ori-small committed:
Rebased and merged changes after Dolev's local changes.
Rebased and merged changes after Dolev's local changes.

(#86c246m16)

ori-small committed:
Added efs permanent flag for file transfer to efs permanent location.
Added efs permanent flag for file transfer to efs permanent location.

(#86c246m16)

ori-small committed:
Fices all hugging face files to support the new flag.
Fices all hugging face files to support the new flag.

(#86c246m16)

ori-small committed:
Reverted changes in flags as they are not needed anymore.
Reverted changes in flags as they are not needed anymore.

(#86c246m16)

ori-small committed:
Added the permanent flag for efs model files transfer.
Added the permanent flag for efs model files transfer.

(#86c246m16)

ori-small committed:
Reverted flag changes according to current sns -> sqs subscriptions and communication.
Reverted flag changes according to current sns -> sqs subscriptions and communication.

(#86c246m16)

ori-small committed:
Changed paths according to the new EFS structure updates.
Changed paths according to the new EFS structure updates.

(#86c246m16)

ori-small committed:
CR fixes: removed permanent flags as it is not used by action track.
CR fixes: removed permanent flags as it is not used by action track.

(#86c246m16)

ori-small committed:
Added version_id to the EFs new hirarchy according to new model schema.
Added version_id to the EFs new hirarchy according to new model schema.

(#86c246m16)

ori-small committed:
Added version_id to huggingface_download_request.py for adapting to the new model hirarchy.
Added version_id to huggingface_download_request.py for adapting to the new model hirarchy.

(#86c246m16)

ori-small committed:
Removed duplicates of asset_id from huggingface_download_request.py.
Removed duplicates of asset_id from huggingface_download_request.py.

(#86c246m16)

ori-small committed:
Removed Model field to become ModelProjection for a single version handling.
Removed Model field to become ModelProjection for a single version handling.

(#86c246m16)

ori-small committed:
Added cuda and python version to model_version, and removed it from model.
Added cuda and python version to model_version, and removed it from model.

(#86c246m16)

ori-small committed:
Added the model id to hugging face tracked message.
Added the model id to hugging face tracked message.

(#86c246m16)

ori-small committed:
Changed asset message to get the asset i d in entity id field.
Changed asset message to get the asset i d in entity id field.

(#86c246m16)

ori-small committed:
Changed flag name to contain asset_name, model_id not needed.
Changed flag name to contain asset_name, model_id not needed.

(#86c730dbp)

ori-small committed:
Removed unnecessary changes + ordered the enum alphabetically.
Removed unnecessary changes + ordered the enum alphabetically.

(#86c730dbp)

yonatankarimish commented on norman_objects/services/hug/huggingface_download_request.py:
```diff
@@ -0,0 +1,42 @@
+    version_id: str
+    asset_id: str
+    asset_name: str
+    huggingface_model_id: str
```
Naming: `huggingface_model_name`

1) ID and Name are used interchangeably in Huggingface
2) Aligns with our property ordering scheme (names after ids)
3) Reflects that the value is closer to a proper name than to an alphanumerical id

yonatankarimish commented on norman_objects/services/hug/huggingface_download_request.py:
```diff
@@ -0,0 +1,42 @@
+    huggingface_model_id: str
+
+    @property
+    def entity_id(self):
```
Design: why do we need this property?

We often use `entity_id` when we need to expose a proper id polymorphically. This class has no subclasses.

yonatankarimish commented on norman_objects/services/hug/huggingface_download_request.py:
```diff
@@ -0,0 +1,42 @@
+        return self.asset_id
+
+    @property
+    def entity_type(self):
```
Design: see above

yonatankarimish commented on norman_objects/services/hug/huggingface_download_request.py:
```diff
@@ -0,0 +1,42 @@
+from norman_objects.shared.status_flags.status_flag_value import StatusFlagValue
+
+
+class HuggingFaceDownloadRequest(NormanBaseModel):
```
Design: It is unclear if this class is standalone or subclasses AssetDownloadRequest. It is probably a good idea to make a choice either way and write the class according to whatever choice you make.

yonatankarimish commented on norman_objects/services/hug/huggingface_download_request.py:
```diff
@@ -0,0 +1,42 @@
+    def entity_type(self):
+        return EntityType.Asset
+
+    def to_base_message(self, flag_value: StatusFlagValue):
```
Design: See comment on class and inheritance.

yonatankarimish commented on norman_objects/services/hug/huggingface_download_request.py:
```diff
@@ -0,0 +1,42 @@
+        status_flag = super().to_status_flag(flag_value)
+        return AssetMessage.base_message(status_flag)
+
+    def to_status_flag(self, flag_value: StatusFlagValue):
```
Design: See comment on class and inheritance.

yonatankarimish commented on norman_objects/services/hug/huggingface_tracked_download.py:
```diff
@@ -0,0 +1,61 @@
+from norman_objects.shared.status_flags.status_flag_value import StatusFlagValue
+
+
+class HuggingFaceTrackedDownload(NormanBaseModel):
```
Design: See comment on `HuggingFaceDownloadRequest` class and inheritance.

Make sure to apply all corrections from that class here as well.

yonatankarimish commented on norman_objects/shared/context/secure_file_context_manager.py:
```diff
@@ -32,13 +32,16 @@ def security_checks(self):
 
         if account_id_segment != self.account_id:
-            raise PermissionError(f"Path account segment {account_id_segment} does not match expected account ID {self.account_id}")
+            raise PermissionError(
```
Style: Remove line break

yonatankarimish commented on norman_objects/shared/context/secure_file_context_manager.py:
```diff
@@ -32,13 +32,16 @@ def security_checks(self):
+            ) = segments[-6:]
         except ValueError:
-            raise ValueError(f"File path {self.path} does not conform to the expected 4-segment structure")
+            raise ValueError(f"File path {self.path} does not conform to the expected 6-segment structure")
```
Question: I thought we discussed the following structure:
```
/mnt/norman
├── /ephemeral
│   └── /{account_id}
│       └──/{model_id}
│           └──/{version_id}
│               ├──/{asset_id}
│               │  └──/logo.asset
│               ├──/{input_id}
│               │   ├──/staging.jpg
│               │   ├──/transcoding.png
│               │   └──/tensor.pt
│               └──/{output_id}
│                   └──/staging
└── /permanent
    └── /{account_id}
        └──/{model_id}
            └──/{version_id}
               ├──/{asset_id}
               │   └──/inference.py
               ├──/{asset_id}
               │   └──requirements.txt
               └──/{asset_id}
                   └──weights.zip
```

With usage as
```
path = f"{mountpoint}/{partition_name}/{account_id}/{model_id}/{version_id}/{entity_id}/{flag_name}.{file_ending}"
```

ori-small replied:
The last time we spoke about this subject we agreed on: f"{mountpoint}/ephemeral/{account_id}/{model_id}/{entity_type}/{object_id}"

entity type here will be Asset/Input/Output
and underneath it all the relevant files will be stored.

yonatankarimish commented on norman_objects/shared/inputs/input_source.py:
```diff
@@ -6,3 +6,5 @@ class InputSource(str, Enum):
     Primitive = "Primitive"
     Stream = "Stream"
+    HuggingFace = "HuggingFace"
+    Git = "Git"
```
Question: Do we use `Git` source at the present?

If not, please remove
Across PR

yonatankarimish commented on norman_objects/shared/model_signatures/receive_format.py:
```diff
@@ -5,3 +5,5 @@ class ReceiveFormat(str, Enum):
     Link = "Link"
     Primitive = "Primitive"
+    HuggingFace = "HuggingFace"
+    Git = "Git"
```
See above

yonatankarimish commented on norman_objects/shared/model_assets/asset_name.py:
```diff
@@ -0,0 +1,8 @@
+from enum import Enum
+
+
+class AssetName(str, Enum):
```
Question: Isn't it about time we open a `/shared/assets` directory for everything asset-related?

If so, move everything there.

yonatankarimish commented on norman_objects/shared/models/model_version.py:
```diff
@@ -33,6 +33,9 @@ class ModelVersion(ModelVersionPreview):
     url: Optional[str] = None
     output_format: OutputFormat
 
+    cuda_version: Optional[str] = None
```
Order: Match order in norman_scripts.

yonatankarimish replied:
Typing: Why are these nullable?

yonatankarimish commented on norman_objects/shared/status_flags/status_flag_name.py:
```diff
@@ -5,9 +5,15 @@ class StatusFlagName(str, Enum):
     File_EFS_Staging = "File_EFS_Staging"
     File_S3_Storage = "File_S3_Storage"
-    File_Image_Build = "File_Image_Build"
+    File_Model_Deploy = "File_Model_Deploy"
```
Naming: See comments on flag naming in `norman_scripts`.

yonatankarimish reviewed (changes requested):
Mostly inline comments. Notice in particular the file system design question.

ori-small committed:
CR fixes: changed huggingface download request to be a standalone class (no need for links and checksums), and changed naming according to cr.
CR fixes: changed huggingface download request to be a standalone class (no need for links and checksums), and changed naming according to cr.

(#86c730dbp)

ori-small committed:
CR fixes: created shared/model_assets folder for better separation.
CR fixes: created shared/model_assets folder for better separation.

(#86c730dbp)

ori-small committed:
CR fixes: Added deployment -> image config to objects as it is used in several places.
CR fixes: Added deployment -> image config to objects as it is used in several places.

(#86c730dbp)
